Support non-tensor inputs and outputs for checkpointed functions.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52422 Support non-tensor inputs and outputs for checkpointed functions.**

As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

In addition to this, I've also updated `gradcheck` to handle non-Tensor outputs appropriately.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

pritamdamania committed:
Support non-tensor inputs and outputs for checkpointed functions.

As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -70,7 +70,7 @@ def forward(ctx, run_function, preserve_rng_state, *args):
                 ctx.had_cuda_in_fwd = True
                 ctx.fwd_gpu_devices, ctx.fwd_gpu_states = get_device_states(*args)
-        ctx.save_for_backward(*args)
+        ctx.inputs = args
```
Would the tensor inputs still be freed on the backward pass here when `retain_graph=False`?
Maybe we could pick out the tensor arguments from inputs and still use `save_for_backward` in those cases and only use ctx to store non-tensor inputs

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -104,16 +104,18 @@ def backward(ctx, *args):
                 " this checkpoint() is not necessary")
         torch.autograd.backward(outputs_with_grad, args_with_grad)
-        grads = tuple(inp.grad if isinstance(inp, torch.Tensor) else inp
+        grads = tuple(inp.grad if isinstance(inp, torch.Tensor) else None
```
Is there a reason we originally returned `inp` for grad if `inp` is not a Tensor? Why does it need to be `None` now? Just curious

gchanan commented:
I'll let @albanD comment but off the top of my head this doesn't look like the right approach.  The issue has to do with checkpoints, but we are now pushing all the complexity into the autograd engine.  Is there some approach we can take where we construct/wrap the function in a smarter way?

gchanan commented:
sorry, I should say "custom function implementation" rather than "autograd engine"

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania87 commented:
> I'll let @albanD comment but off the top of my head there might be a nicer approach. The issue has to do with checkpoints, but we are now pushing all the complexity into the custom function. Is there some approach we can take where we construct/wrap the function in a smarter way?

I guess the PR and the issue mention checkpointing, but I thought of fixing this in general for custom functions as well. Pipeline parallelism uses a bunch of custom functions and it would be nice if this was supported in custom functions as well instead of just checkpointing.

gchanan commented:
I see -- could you point to use cases outside of checkpointing?

pritamdamania87 commented:
> I see -- could you point to use cases outside of checkpointing?

This is the use case in pipeline parallelism: https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/checkpoint.py#L249. It uses its own custom autograd function for checkpointing since it reduces GPU idle time. Full explanation can be found at the top of the checkpoint.py file.

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

soulitzer commented on test/test_utils.py:
```diff
@@ -269,6 +269,47 @@ def run_fn(tensor1, tensor2):
+        t1_grad = t1.grad
+        t2_grad = t2.grad
+
+        # Rest grads, run without checkpoint and validate we receive same grads.
```
nit: Rest -> Reset

malfet replied:
To repeat @soulitzer's suggestion
```suggestion
        # Reset grads, run without checkpoint and validate we receive same grads.
```

soulitzer commented:
Maybe add a test for when none of the inputs/outputs are tensors? You probably also want to fix the clang-tidy failure, but LGTM otherwise. 

I'll let @albanD give the final approval since he mentioned he wanted to take a quick look once he's back.

albanD commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -102,8 +102,14 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+  uint32_t num_tensors = 0;
   for (auto i = 0; i < num_outputs; ++i) {
     Variable var = raw_outputs[i];
+    if (!var.defined()) {
```
What if the user returned an undefined Tensor on which he wants history to be tracked?

albanD replied:
The thing is that this code is in common for both python and c++
But I could see users trying to track gradients for their custom objects and using undefined Tensors as a "cheap" object to give to the autograd. Not that I would recommend doing that :p 

I think that if you only want to make the python version work for checkpoint, it will be simpler for sure to do a wrapper on checkpoint.

Otherwise, I think we will have to update this function to take list of optional Tensor to be able to properly mark the outputs that should not be considered and/or use mark_non_differentiable. But that might have more perf implications...

pritamdamania87 replied:
> But I could see users trying to track gradients for their custom objects and using undefined Tensors as a "cheap" object to give to the autograd. Not that I would recommend doing that :p

Hmm I'm still not sure how an undefined Tensor can be used to track autograd history? If the output of a function is an undefined tensor, can't it be ignored for autograd purposes (as we do here)? 

> I think that if you only want to make the python version work for checkpoint, it will be simpler for sure to do a wrapper on checkpoint.

I guess the goal here is to support this for python custom functions in general and not just checkpoint. In particular classes like these: https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/checkpoint.py#L249.

albanD replied:
My question is that if we only want to support that for python function, is there a way to make this change python only? That will make things easier to review and less likely to break the c++ Function in a subtle way.

albanD commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -133,13 +139,14 @@ variable_list _wrap_outputs(const variable_list &input_vars,
   if (num_diff_outputs > 1) {
     for (auto& var: outputs) {
-      if (var.is_view()) {
+      if (var.defined() && var.is_view()) {
```
Why do you need to check if var is defined here? You bail out at the beginning of the loop in the other case no?

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -366,8 +366,8 @@ static void _wrap_outputs(const std::shared_ptr<PyNode>& cdata, THPFunction *sel
-    throw TypeError("%s.forward: expected Tensor or tuple of Tensor (got %s) for return value %d",
-        Py_TYPE(self)->tp_name, Py_TYPE(obj)->tp_name, i);
+    // Return undefined variable as a placeholder.
+    return Variable();
```
Can't this be confused with a user provided undefined Tensor?

albanD commented on test/test_utils.py:
```diff
@@ -269,6 +269,47 @@ def run_fn(tensor1, tensor2):
+        t1.grad = None
+        t2.grad = None
+        res = foo(t1, t2, scale, t3)
+        res[1].sum().backward()
```
nit: you can do all of that in a single backward call

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -82,7 +97,17 @@ def backward(ctx, *args):
+
+        # Fill in inputs with appropriate saved tensors.
+        tensors_idx = 0
+        for idx in tensor_indices:
```
you can use enumerate to remove the need for tensors_idx

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -104,17 +129,20 @@ def backward(ctx, *args):
         # run backward() with only tensor that requires grad
         outputs_with_grad = []
         args_with_grad = []
+        args_index = 0
```
nit: tensor_args_index

albanD reviewed (commented):
If we go with this, tests for both the python and c++ version of the custom Function are missing as well as doc update for both custom Functions and checkpoint.

malfet commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -113,9 +119,9 @@ variable_list _wrap_outputs(const variable_list &input_vars,
     if (cdata) {
       auto output_nr = cdata->add_input_metadata(var);
-      AT_ASSERT(i == (int)output_nr);
+      AT_ASSERT(num_tensors == (int)output_nr);
```
Both num_tensors and output_nr are `unit32_t`, please remove the cast
```suggestion
      AT_ASSERT(num_tensors == output_nr);
```

pritamdamania87 commented:
> Maybe add a test for when none of the inputs/outputs are tensors?

Is this a valid use case from a user point of view though? If there are no tensors to a module, there is no need to run backward through the module and I'm not sure if a custom autograd function makes sense? For example checkpointing today doesn't allow this: https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L24

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
> If we go with this, tests for both the python and c++ version of the custom Function are missing as well as doc update for both custom Functions and checkpoint.

I made these changes in the latest update to the PR for python custom functions. I didn't intend to make these changes for the cpp custom functions as well since those actually still have strict requirements for outputs being Variables. I was wondering if the scope can only be python custom function support? The changes I made in custom_function.cpp were only to ensure python custom functions can support a variety of datatypes.

soulitzer commented:
> > Maybe add a test for when none of the inputs/outputs are tensors?
> 
> Is this a valid use case from a user point of view though? If there are no tensors to a module, there is no need to run backward through the module and I'm not sure if a custom autograd function makes sense? For example checkpointing today doesn't allow this: https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L24

Probably not. If we don't expect users to do this, we could add a check similar to the one you linked, but for custom functions in general.

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania87 commented:
@albanD I updated the PR to get rid of any changes to C++ custom functions and kept the changes python only.

@soulitzer Added tests for none of the inputs/outputs being tensors. Looks like checkpoint supports this and only raises a warning.

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -362,30 +362,40 @@ static void _wrap_outputs(const std::shared_ptr<PyNode>& cdata, THPFunction *sel
+    PyObject* obj = PyTuple_GET_ITEM(raw_output, i);
+    bool is_tensor = THPVariable_Check(obj);
+
+    if (is_executable && is_tensor) {
```
nit: this could be moved inside the if statement below

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -557,11 +567,14 @@ static void _trace_post_record(
-      jit::tracer::setValueTrace(var->cdata, value);
+    PyObject* obj = PyTuple_GET_ITEM(output_objects, i);
+    if (THPVariable_Check(obj)) {
+      auto var = (THPVariable*)PyTuple_GET_ITEM(output_objects, i);
```
you can re-used obj here.

albanD commented on torch/autograd/function.py:
```diff
@@ -173,7 +173,7 @@ def forward(ctx: Any, *args: Any, **kwargs: Any) -> Any:
         number of arguments (tensors or other types).
 
-        The context can be used to store tensors that can be then retrieved
+        The context can be used to store tensors/inputs that can be then retrieved
```
Why can it be used to saved inputs?
The extending doc explicitly says that you should use save_for_backward for inputs: https://pytorch.org/docs/stable/notes/extending.html

albanD replied:
I think this sentence means that you can do `ctx.foo =` to save anything that is not input/output.
But this is most likely not clear indeed.

albanD replied:
I am still not sure that this update makes sense. "tensors/inputs" is not really accurate because you can save pretty much everything you want, including intermediary result and now non-tensor outputs as well.

albanD reviewed (commented):
Looks much better thanks!

I don't see any mention of what the backward() function will get as input when the forward returns non-Tensor objects.
Are they going to be None gradients? (just like we do for the grad_input)

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
> I don't see any mention of what the backward() function will get as input when the forward returns non-Tensor objects.
Are they going to be None gradients? (just like we do for the grad_input)

Updated the docs to include this as well.

albanD commented on torch/autograd/function.py:
```diff
@@ -186,10 +186,13 @@ def backward(ctx: Any, *grad_outputs: Any) -> Any:
-        tensors, as there were inputs to :func:`forward`. Each argument is the
-        gradient w.r.t the given output, and each returned value should be the
-        gradient w.r.t. the corresponding input.
+        as many tensor outputs as the :func:`forward` returned (non tensor
```
If we just ignore them, then what about having more constraint here that you must provide all Tensor outputs first, then as many non-Tensor objects that will be ignored?
That way the indexing of the Tensors properly match.

Beyond that, I am not 100% convinced about just ignoring these. It will make the forward/backward signature non-symetric which I am not sure I like..
cc @gchanan do you have an opinion on this?

pritamdamania87 replied:
> If we just ignore them, then what about having more constraint here that you must provide all Tensor outputs first, then as many non-Tensor objects that will be ignored?
That way the indexing of the Tensors properly match.

I'm not sure if we need to worry about Tensor indexing. From an autograd point of view it only sees that the autograd function outputs tensors and no other values. 

> Beyond that, I am not 100% convinced about just ignoring these. It will make the forward/backward signature non-symetric which I am not sure I like..

I don't think the forward and backward signatures were symmetric earlier either right? You can write a forward function which takes as input 5 tensors and outputs only 3 tensors. In that case, your backward function would get as input 3 tensors, but it would return 5 tensors (for grads of the inputs).

pritamdamania87 replied:
@albanD @gchanan Wondering if you had any more thoughts regarding this specific issue? Would like to get this change in soon to unblock some of our work on pipeline parallelism to support custom types :)

albanD replied:
I think we definitely want to keep this matching between the number of outputs of the forward and the number of inputs in the backward.

albanD replied:
And having all of them contiguous at the beginning doesn't feel necessary indeed, we can drop this idea.

pritamdamania87 replied:
I updated the PR to make this symmetric, but had to touch `custom_function.cpp` for this since that controls how the autograd function metadata is setup. Although, I did it in a way such that it only impacts python custom functions and not cpp custom functions.

soulitzer commented:
Can we add a test for gradcheck? I don't think gradcheck currently expects outputs to be non tensors

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -411,8 +411,8 @@ def fn(input):
 
-    for i, o in enumerate(output):
-        if not o.requires_grad:
+    for i, o in enumerate(func_out):
```
Why change this? The filtering above should be already removing all the non-Tensor outputs?

albanD replied:
I am not sure about these changes.
Can you add some tests to make sure it does the right thing if you keep these?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -289,7 +289,7 @@ def _as_tuple(x):
 
 def _differentiable_outputs(x):
-    return tuple(o for o in _as_tuple(x) if o.requires_grad)
+    return tuple(o for o in _as_tuple(x) if torch.is_tensor(o) and o.requires_grad)
```
Why do you need to change gradcheck here?
I feel like this should be in a separate PR because that may have some unexpected impact as well....

pritamdamania87 replied:
I added gradcheck as per @soulitzer's suggestion. I guess it makes sense to fix `gradcheck` as well as part of this PR since without this, someone could call `gradcheck` on the output of a custom autograd function and it would fail.

albanD replied:
Right but is this exercised / tested?

pritamdamania87 replied:
Yes, I've added a test for this. See last line of `test_custom_function_non_tensor_inputs_outputs`.

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -362,30 +362,38 @@ static void _wrap_outputs(const std::shared_ptr<PyNode>& cdata, THPFunction *sel
+    // Keep the non-tensor outputs as is.
+    if (!THPVariable_Check(obj)) {
+      if (is_executable) {
+        self->output_info.emplace_back(wrapped_outputs[tensor_output_idx]);
```
This should be in the else statement no?

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -557,11 +565,14 @@ static void _trace_post_record(
-      value->inferTypeFrom(var->cdata);
-      jit::tracer::setValueTrace(var->cdata, value);
+    PyObject* obj = PyTuple_GET_ITEM(output_objects, i);
+    if (THPVariable_Check(obj)) {
```
Is it ok for the jit to not trace some values? Won't that create wrong jit IR?

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -82,7 +97,15 @@ def backward(ctx, *args):
                 " is passed to .backward(). Please use .backward() and do not pass its `inputs`"
                 " argument.")
-        inputs = ctx.saved_tensors
+        # Copy the list to avoid modifying original list.
```
The original object is a tuple no? not a list

albanD replied:
Ho sorry I though this was coming from the saved_tensors

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -94,7 +117,7 @@ def backward(ctx, *args):
                 if ctx.had_cuda_in_fwd:
                     set_device_states(ctx.fwd_gpu_devices, ctx.fwd_gpu_states)
-            detached_inputs = detach_variable(inputs)
+            detached_inputs = detach_variable(tuple(inputs))
```
This function is just used here no? Maybe just update the function to take a list as input instead of a tuple?

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -104,17 +127,20 @@ def backward(ctx, *args):
+            if torch.is_tensor(outputs[i]) and outputs[i].requires_grad:
                 outputs_with_grad.append(outputs[i])
-                args_with_grad.append(args[i])
+                args_with_grad.append(args[tensor_args_index])
```
This indexing feels bad. Following the update to include None in the grad_output, that will simplify this code a lot!

albanD reviewed (commented):
I updated my answer below for handling of the non-Tensor outputs.
Also added a few comments inline where the code is a bit surprising.

I haven't looked at the test yet, will do once the grad_output update is in.

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -557,11 +567,19 @@ static void _trace_post_record(
+        jit::tracer::setValueTrace(var->cdata, value);
+      }
+    } else {
+      // Regular python type.
+      Value* value = node->outputs()[i];
+      value->setType(PyObjectType::create());
+      jit::tracer::setValueTrace(torch::jit::toIValue(py::cast(obj), PyObjectType::create()), value);
```
@suo I was wondering if you had some suggestions on how to test JIT is doing the right thing here :)

suo replied:
I don't have any special insight except for to do some trace checking tests on representative functions (similar to the ones in `test_jit.py`)

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

In addition to this, I've also updated `gradcheck` to handle non-Tensor outputs appropriately.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -362,30 +363,39 @@ static void _wrap_outputs(const std::shared_ptr<PyNode>& cdata, THPFunction *sel
+        self->output_info.emplace_back();
+      }
+      Py_INCREF(obj);
+      PyTuple_SET_ITEM(outputs, i, obj);
```
double check that the capital version of this macro is not leaking memory

albanD commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -9,18 +9,26 @@ VariableInfo::VariableInfo(const Variable& var)
   , size(var.sizes().vec())
-  , requires_grad(var.requires_grad()) {
+  , requires_grad(var.requires_grad())
+  , is_null(false) {
```
nit is_empty ?

albanD reviewed (commented):
- Change to custom function is good. just nit for naming
- checkpoint doc should be updated to tell what is accepted as input/output and what won't work (nested structures)
- gradcheck change should be reverted as we don't want to support non-tensor output there
- Add a wrapper function to be able to test the new feature with gradcheck

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

In addition to this, I've also updated `gradcheck` to handle non-Tensor outputs appropriately.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -135,6 +159,12 @@ def checkpoint(function, *args, **kwargs):
+    The output of :attr:`function` can contain non-Tensor values and gradient
+    recording is only performed for the Tensor values. Note that if the output
+    consists of nested structures (ex: custom objects, lists, dicts etc.)
+    consisting of Tensors, these Tensors nested in custom structures would not
```
nit would -> will

albanD reviewed (approved):
Checked the points from the comments above.

LGTM

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

In addition to this, I've also updated `gradcheck` to handle non-Tensor outputs appropriately.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-tensor inputs and outputs for checkpointed functions."


As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

In addition to this, I've also updated `gradcheck` to handle non-Tensor outputs appropriately.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

[ghstack-poisoned]
